### PR TITLE
Glm4 mtp optimizations

### PR DESCRIPTION
**Pull request by SamuelOliveirads**
I've created this draft to share my findings on what to fix or improve to make MTP usable. Currently, MTP's output quality is good, but its performance is worse than not using it at all. Therefore, it's not enough to be on par with the baseline; we need to be faster.

 My initial plan is to find areas for improvement. It's not necessary to implement everything at once, but some of these should be on our radar for the future. They are:
 1) Graph reuse
 2) `llama_context::decode` calls
 3) Multi-token drafts

 There are likely more things to improve, but for now, I find these to be the most impactful. Below are my thoughts on each:

 **1) Graph Reuse:** The baseline implementation always reuses the graph. The process is simple: it stores the graph, and in the next call to `llama_context::process_ubatch`, it checks if the stored graph can be reused. If not, it's deleted and the new one is stored. This works well after the first token is generated, as subsequent graphs are identical. The main bottleneck isn't calling `llama_model::build_graph` constantly, but rather `ggml_backend_sched_alloc_graph`, which has to allocate and compute resources for the backend.

 The first fix was simple: just store one graph. In this case, the main model's token generation graph, which is one of the most expensive, will always be reused. On my machine, this gave an uplift of 13.8% for small prompts.

 **Current state: Halted.**

 After that, I tried to store the graph for every operation, or at least the ones that didn't involve the KV cache. By applying `llm_graph_context::cb` to certain layers, I could store and reuse the graph, and I was able to compile and test this using only the CPU backend. However, I was unable to get it working with the offload policy. In theory, the `cb` function should handle that, but something else seems to be preventing specifically the allocation and computation. Is it mixing the offload policies of the main model and the MTP? This needs a deeper investigation, and I lack the proper knowledge in this area, so I'm setting it aside for now.

 **2) `decode` calls:** MTP was successfully implemented inside `decode`, but it uses the old logic where each operation requires an expensive function call. Here is a comparison of how many calls we make in different scenarios:

 *   **LLM - Normal:**
     *   Loop 1: Prompt + Generation = 1 call
     *   Loop 2: Token generation = 1 call

 *   **Draft Model:**
     *   Loop 1: Prompt + Generation -> Draft generation -> Main model validation = 3 calls
     *   Loop 2: Token generation -> Draft generation -> Main model validation = 3 calls

 *   **MTP (Current Slow Implementation):**
     *   Loop 1: Prompt + Generation -> MTP warmup -> MTP draft -> Main model validation -> MTP KV update = 5 calls
     *   Loop 2: Token generation -> MTP draft -> Main model validation -> MTP KV update = 4 calls

 One way to make MTP more usable is to match the number of calls of a typical draft model. To do that, it's necessary to combine the KV cache update and the draft generation into a single call.

 **Current state: In progress.**

 I successfully merged the KV cache update with the draft generation. This required creating a custom batch and sinfo, and changing some logic regarding the embeddings and hidden states necessary for the MTP to work. The version in this branch works in terms of output, meaning it's not breaking quality. However, the draft acceptance rate has dropped to around 25%. I believe this happens because while the first step (KV update) works using the correct hidden state from the main model, the subsequent operation (draft) is using a new hidden state generated by the MTP itself during the update. I still need to confirm this theory and apply a fix to hopefully see the acceptance rate rise back to its previous level.

 One last thing: this change will still require a separate warmup call on the first interaction, but this is less impactful than merging the update and draft steps. To merge the warmup step, it would be necessary to track the sinfo to know when the prompt processing has finished its last batch, and then insert a new slot for the draft token.

 **3) Multi-token drafts:** We discussed this in another PR. The problem was that for each new draft token, the MTP's KV cache needed to be updated, which was painful to do before. Now that we are using the `decode` function, it's more feasible. If the unified update/draft implementation works, we could simply increase the batch and sinfo size to make the model draft more tokens.

 These are some of my ideas. I'd appreciate any insights you might have on how to better handle some of these things, or even new ideas for improvements that I haven't spotted here.

### CODE REVIEW

**Comment by SamuelOliveirads** [comment]
Hi @F1LM1,

I've implemented the logic mentioned regarding the `llama_context::decode` calls. It makes the flow much simpler and easier to understand. We now essentially call `decode` three times: once for the main model (prompt + token), once for the MTP draft, and finally for validation. The KV cache updates have also been merged into these calls to minimize overhead.

One issue I'm currently investigating is a potential bug with position handling in large prompts. The model sometimes ignores the stop token for "thinking" or repeats the thought process in the final reply.

In terms of results, I didn't observe a major boost in tokens/s yet, but the code is now much cleaner to maintain and sets a better foundation for applying optimizations like graph reuse and compute management.

I also wanted to touch base regarding the future of this MTP implementation. I shared my roadmap in the original PR, but since you are the owner, I'd like to know your plans/availability.

Thanks for your time!

**Comment by F1LM1** [comment]
Hey @SamuelOliveirads,

I've been following along with the PRs even though I've been quiet these past few weeks. Been busier + I feel that I have little experience on the optimization front so haven't been able to contribute much, sadly. Outside of random bugs/lower-level optimizations like was just found in https://github.com/ggml-org/llama.cpp/pull/15225#issuecomment-3566589474, the two things that stand out to me:

1. In the original PR I noted an issue where we alternate between main model calls with draft and main model calls without drafting. This is not an MTP-specific issue but it'll especially hurt MTP performance versus typical draft models because we're only drafting one token at once without multi-token. I'm referring to this:
    ```
    Loop 2: Token generation -> Draft generation -> Main model validation = 3 calls
    ```
    which could really be simplified to two calls: Draft generation -> Main model validation -> Draft generation -> Main model validation, etc. after the first "standard" token is generated, since model validation itself produces a previous main-model-validated token that can be plugged into the MTP head. Think this is low-hanging fruit, it could be as simple as using a boolean to make sure the main decode loop (most of lines 3474 to 3587) only run once. With MTP draft acceptance rates as high as they are, this should provide a decent speedup during generation, and this will scale as we continue to optimize MTP overhead.

2. If we want to hit the ambitious performance speedups claimed by MTP elsewhere we will have to get started on multi-token.

As for high-level questions about this PR's future, I'm not sure what standard we would have to hit to get it pushed through to the main llama.cpp branch. Noticed you asked ggerganov the same in the original PR. It's a question I would love to know the answer to as well. But I'd guess that if we can get this to the point where it's providing meaningful speedup over no speculative draft, it will be worth merging the original PR, at which point further optimizations can go in their own PRs.

I'm hoping [7c4b2c1](https://github.com/F1LM1/llama.cpp/pull/4/commits/7c4b2c15708173b9c228dabe3ab82c50032e9afb) gets overhead low enough that combined with my first suggestion here and maybe some more minor optimizations, it wouldn't be a reach to hit something like 30-40% speedup over no MTP, which I think would be a reasonable checkpoint to push to get the original PR done with. Does that seem reasonable to you?

**Comment by SamuelOliveirads** [comment]
> I've been following along with the PRs even though I've been quiet these past few weeks. Been busier + I feel that I have little experience on the optimization front so haven't been able to contribute much, sadly.

No worries! I'm also digging deeper into this area to identify what can be fixed and improved.

> 1. In the original PR I noted an issue where we alternate between main model calls with draft and main model calls without drafting. This is not an MTP-specific issue but it'll especially hurt MTP performance versus typical draft models because we're only drafting one token at once without multi-token. I'm referring to this:
>    ```
>    Loop 2: Token generation -> Draft generation -> Main model validation = 3 calls
>    ``` 
>      
>    which could really be simplified to two calls: Draft generation -> Main model validation -> Draft generation -> Main model validation, etc. after the first "standard" token is generated, since model validation itself produces a previous main-model-validated token that can be plugged into the MTP head. Think this is low-hanging fruit, it could be as simple as using a boolean to make sure the main decode loop (most of lines 3474 to 3587) only run once. With MTP draft acceptance rates as high as they are, this should provide a decent speedup during generation, and this will scale as we continue to optimize MTP overhead.

I was initially planning to address this in a future PR, but it feels like it is necessary now to make MTP truly usable. I believe that, apart from a skip flag, we need to track tokens and potentially other metrics. I'll take some time to better understand the consequences of running without the distinct main decode step.

> 2. If we want to hit the ambitious performance speedups claimed by MTP elsewhere we will have to get started on multi-token.

I have a plan for that. For each new drafted token, I'll need to collect and store the hidden state of the previous MTP drafted token. The MTP graph will also need to output the hidden state for storage. Once that's in place, we can implement a loop inside `mtp_speculative_gen_draft` to handle each token in the batch.

> I'm hoping [7c4b2c1](https://github.com/F1LM1/llama.cpp/pull/4/commits/7c4b2c15708173b9c228dabe3ab82c50032e9afb) gets overhead low enough that combined with my first suggestion here and maybe some more minor optimizations, it wouldn't be a reach to hit something like 30-40% speedup over no MTP, which I think would be a reasonable checkpoint to push to get the original PR done with. Does that seem reasonable to you?

I hope so too. It's good to have a concrete goal. I believe your two suggestions, combined with saving one of the model graphs (preferably the larger validation one), will yield that speedup. Even if MTP doesn't quite hit that target, we can open it up to testers to see if different backends, configurations, and use cases provide better performance.

**Comment by wishstudio** [comment]
I think point 1 is probably the best next step. A hack should be relatively easy (I'm still wrapping my head on this). But a proper implementation may involve redesigning of the relevant code block.

Point 2 looks good on paper, but longer drafts result in worse acceptance rates so I'm a bit skeptical of how much speedup this can get. There are a many research work on optimizing this, like the techniques used in EAGLE papers. They are however quite complicated to implement.

Speaking of overheads I can still spot some small ones, but profiling shows the total room of overhead optimizations is like ~15%, so not a big deal.

Right now I'm getting like ~10% speedup over without MTP with `--cpu-moe`. My napkin math shows 30-40% speedup is probably the ceiling, with all these goals implemented and overheads fixed. I believe the claimed big speed-ups can only be achieved in pure GPU workloads because batch size 1 is very inefficient in GPUs. But testing this is outreach for me.

**Comment by SamuelOliveirads** [comment]
> Point 2 looks good on paper, but longer drafts result in worse acceptance rates so I'm a bit skeptical of how much speedup this can get. There are a many research work on optimizing this, like the techniques used in EAGLE papers. They are however quite complicated to implement.

I don't know too much about other research on MTP performance, but I would love to read some. Looking at the [example ](https://nvidia.github.io/TensorRT-LLM/blogs/tech_blog/blog1_Pushing_Latency_Boundaries_Optimizing_DeepSeek-R1_Performance_on_NVIDIA_B200_GPUs.html#autoregressive-mtp-layers) from the original PR regarding Nvidia's implementation, a single-token draft yielded a speedup of roughly 78%, going up to 128% for 3 drafts. We cannot expect the same gains, as they probably used only GPUs, and we don't know what the acceptance rate was. Most importantly, this is a first implementation that definitely will need more polish, but the potential is good enough to try.

**Comment by wishstudio** [comment]
I've now read most of this implementation and are feeling it's better to "revert" to separate speculative model (regarding @F1LM1 's comments in #2). 

Conceptually although the main model and MTP heads are presented as a single unified model, I found no problem to think of them as different models. The MTP model only needs the hidden state and token outputs of the main model and work on its own. Compared to conventional speculation decoding, the only extra hidden state input needs not much different handling than the token input. It also has its own KV cache management that does not depend on anything from the main model.

Down to implementation, I feel the underlying codebase is much more compatible with defining MTP as a separate model. That's how current speculative decoding works, after all. We should also be able to reuse some common facilities. From an optimization perspective, the current unified approach in this PR is also suboptimal: the MTP KV update graph is merged into the main model, but it's still separate kernel calls. `build_mtp_draft_graph` and `build_mtp_update_graph` are essentially building the same graph, just working on different part of the inputs. Kernel-wise currently the entire process is still the `main model -> update draft kv -> draft model` trio. But the `update draft kv` and `draft model` series of kernel calls can be merged into one. It simply needs to roll back speculations and "catch up" to main model's outputs in a single batch, similar to what current speculative decoding is doing. Separating them into two models will also eliminate the need of `MTP_OP_*` handling and make the code even more cleaner.

The only drawback to this approach is we lost the ability to easily batch speculative decoding across multiple slots. Because now different slots could have different length of tokens to run the draft model due to different validation results. But since a) the current speculation decoding does not support slot batching at all b) we can forcibly batching them by using masks and I think the extra wasted computation is negligible, I don't think this is a problem.

Regarding reducing the model calls to 2, to my current understanding of the relevant code I believe there is no major roadblocks and should be doable with a few additional `if`s to skip main model call as @F1LM1 suggested. But I guess this should better be another PR after we get this one merged because this is quite independent to the MTP implementation and could also help conventional speculative decoding.

**Comment by SamuelOliveirads** [comment]
> build_mtp_draft_graph and build_mtp_update_graph are essentially building the same graph, just working on different part of the inputs.

Indeed, the idea was to remain the same; after all, the MTP still needs to update its cache to generate new drafts. My objective in merging the graphs was simply to avoid many `llama_decode` calls. It was bad to handle for the scheduler and `graph_reuse` as we often need to generate around 4 graphs per interaction.

The current implementation also has two possible problems: the `graph_reuse` that was previously partly activated and now I fully deactivated. In previous tests, this gave me even a slight boost over the original PR of around 13.8%. The other is the possibility of bugs, although I didn't find anything wrong in my tests.

> But I guess this should better be another PR after we get this one merged because this is quite independent to the MTP implementation and could also help conventional speculative decoding.

This is kind of the main issue for now. I can simply create a PR for the fix with `llama_get_logits_ith`, reactivate the partial `graph_reuse` and better usability for the user to choose when to use MTP or not. Then @F1LM1 will need to open his original PR and we hope for someone to review it. Otherwise, we should probably only find enough traction if we actually get better token speed with MTP than without (which is the idea of this heavy PR, and I agree that it's a lot to change here).

**Review by RatStar811 — APPROVED** [review]
